core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1816,8 +1292,12 @@ class Log(@volatile private var _dir: File,
    */
   private def deleteOldSegments(predicate: (LogSegment, Option[LogSegment]) => Boolean,
                                 reason: SegmentDeletionReason): Int = {
+    def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment], logEndOffset: Long): Boolean = {
+      highWatermark >= nextSegmentOpt.map(_.baseOffset).getOrElse(logEndOffset) &&
```

Comment:
Hmm, why do we need to wrap `predicate` with an additional condition?

Reply:
this is to accomodate for the hwm check that was previously happening in `Log#deletableSegments` in this [line](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/Log.scala#L1872). The `deletableSegments` method has now moved to `LocalLog`, so we piggyback on the predicate to additionally attach the hwm check.